### PR TITLE
Refactor `CONFIG_SPE` check on Linux/powerpc

### PR DESCRIPTION
**Pull request by Low-power**
### Motivation and Context
Commit 5401472 adds a check to call `enable_kernel_spe` and `disable_kernel_spe` only if `CONFIG_SPE` is defined. This introduces redundant implementations of `kfpu_begin()` and `kfpu_end()`.

### Description
Refactor this check in a way similar to what `CONFIG_ALTIVEC` and `CONFIG_VSX` are checked in #14591, in order to remove redundant codes.

### How Has This Been Tested?
Tested on my POWER7-based machine without `CONFIG_SPE`. May need more testings on machines with `CONFIG_SPE`.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Review by mcmilk — APPROVED** [review]

**Comment by Low-power** [comment]
Hold on!

When testing with `CONFIG_SPE=y`, I found an severe issue in this change.

Working on fixes.

**Comment by Low-power** [comment]
The semicolons after the functions calls are missing from #14591. I have no idea how it didn't trigger a build failure when I testing #14591 for myself.

Has added a separated commit (6b8195678e89072ed3ff36759a06f905ae18663a) to fix it.


> it was also wrong before.

You are right.

**Comment by mcmilk** [comment]
I like this code snippet more:

```
#define kfpu_begin()                            \
        {                                       \
                preempt_disable();              \
#ifdef  CONFIG_ALTIVEC
                enable_kernel_altivec();
#endif
#ifdef  CONFIG_VSX
                enable_kernel_vsx();
#endif
#ifdef CONFIG_SPE
                enable_kernel_spe();
#endif
        }
#define kfpu_end()                              \
        {                                       \
#ifdef CONFIG_SPE
                disable_kernel_spe();           \
#endif
#ifdef  CONFIG_VSX
                disable_kernel_vsx();           \
#endif
#ifdef  CONFIG_ALTIVEC
                disable_kernel_altivec();       \
#endif
                preempt_enable();               \
        }
#endif

```

**Comment by Low-power** [comment]
> I like this code snippet more:

But `#if` didn't work with continued lines.

**Comment by Low-power** [comment]
> Can we take the `;` to the define like this:

This should also be ok, I think.

**Review by behlendorf — APPROVED** [review]

File: include/os/linux/kernel/linux/simd_powerpc.h [hunk 1]
```diff
@@ -80,40 +80,32 @@
...
-	{					\
-		disable_kernel_spe();		\
-		DISABLE_KERNEL_VSX		\
-		DISABLE_KERNEL_ALTIVEC		\
-		preempt_enable();		\
-	}
-#else /* CONFIG_SPE */
+#ifdef	CONFIG_SPE
+#define	ENABLE_KERNEL_SPE	enable_kernel_spe()
+#define	DISABLE_KERNEL_SPE	disable_kernel_spe()
```
Comment (mcmilk) [inline include/os/linux/kernel/linux/simd_powerpc.h:85]: From:
```
#define	ENABLE_KERNEL_SPE	enable_kernel_spe()
#define	DISABLE_KERNEL_SPE	disable_kernel_spe()
```

To:
```
#define	ENABLE_KERNEL_SPE	enable_kernel_spe();
#define	DISABLE_KERNEL_SPE	disable_kernel_spe();
```

Then I will approve ;)

File: include/os/linux/kernel/linux/simd_powerpc.h [hunk 2]
```diff
@@ -80,40 +80,32 @@
...
+#define	DISABLE_KERNEL_SPE
+#endif
 #define	kfpu_begin()				\
 	{					\
 		preempt_disable();		\
-		ENABLE_KERNEL_ALTIVEC		\
-		ENABLE_KERNEL_VSX		\
+		ENABLE_KERNEL_ALTIVEC;		\
+		ENABLE_KERNEL_VSX;		\
+		ENABLE_KERNEL_SPE;		\
```
Comment (mcmilk) [inline include/os/linux/kernel/linux/simd_powerpc.h:95]: And here the other way:
`ENABLE_KERNEL_SPE;` --> `ENABLE_KERNEL_SPE`

**Review by mcmilk — CHANGES_REQUESTED** [review]

File: include/os/linux/kernel/linux/simd_powerpc.h (hunk 1 above)
Comment (behlendorf) [reply include/os/linux/kernel/linux/simd_powerpc.h:85]: Why the preference for including the `;` in the #define?  This seems like it should be purely cosmetic, and our usual convention is to not hide it in the macro.  Though I don't feel strongly about it.

**Comment by Low-power** [comment]
Before I going to amend my commits, I would like to ask, are you sure this is better?

I mean, by searching the source tree, the only occurrences of `#define` with trailing `;` are some pseudo instructions for assembler:
```
whr@debian:~/src/zfs$ grep -E '#[[:space:]]*define[[:space:]]+[A-Z0-9_]+[[:space:]].+;$' --include "*.[ch]" -r *
include/os/linux/spl/sys/ia32/asm_linkage.h:#define	D16	data16;
include/os/linux/spl/sys/ia32/asm_linkage.h:#define	A16	addr16;
include/os/linux/spl/sys/ia32/asm_linkage.h:#define	D16	.byte	0x66;
include/os/linux/spl/sys/ia32/asm_linkage.h:#define	A16	.byte	0x67;
include/os/freebsd/spl/sys/ia32/asm_linkage.h:#define	D16	data16;
include/os/freebsd/spl/sys/ia32/asm_linkage.h:#define	A16	addr16;
include/os/freebsd/spl/sys/ia32/asm_linkage.h:#define	D16	.byte	0x66;
include/os/freebsd/spl/sys/ia32/asm_linkage.h:#define	A16	.byte	0x67;
lib/libspl/include/os/linux/sys/ia32/asm_linkage.h:#define	D16	data16;
lib/libspl/include/os/linux/sys/ia32/asm_linkage.h:#define	A16	addr16;
lib/libspl/include/os/linux/sys/ia32/asm_linkage.h:#define	D16	.byte	0x66;
lib/libspl/include/os/linux/sys/ia32/asm_linkage.h:#define	A16	.byte	0x67;
lib/libspl/include/os/freebsd/sys/ia32/asm_linkage.h:#define	D16	data16;
lib/libspl/include/os/freebsd/sys/ia32/asm_linkage.h:#define	A16	addr16;
lib/libspl/include/os/freebsd/sys/ia32/asm_linkage.h:#define	D16	.byte	0x66;
lib/libspl/include/os/freebsd/sys/ia32/asm_linkage.h:#define	A16	.byte	0x67;
```

**Comment by Low-power** [comment]
Tested with a few different configurations of Linux 5.15.96.

File: include/os/linux/kernel/linux/simd_powerpc.h (hunk 1 above)
Comment (mcmilk) [reply include/os/linux/kernel/linux/simd_powerpc.h]: Yes, it's just cosmetic. My variant just defines it `all away` - which seems nicer to me.
But when there is a usual convention which does does not hide the `;` - then we should follow this convention.

@Low-power - sorry for the noise ;-)

**Review by mcmilk — APPROVED** [review]

**Comment by mcmilk** [comment]
> Before I going to amend my commits, I would like to ask, are you sure this is better?

Sorry ... if you want, you can rebase to the first variant. Both variants are fine and fix the ppc64 code path.

**Comment by Low-power** [comment]
Well, if I change the **grep(1)** command to
```
grep -E '#[[:space:]]*define[[:space:]]+[A-Za-z0-9_(,)]+[[:space:]].+;$' --include "*.[ch]" -r *
```
it will give some more results.
While some results are indeed questionable (will producing redundant `;`), some other such as `CALLB_CPR_ASSERT` in `include/os/freebsd/spl/sys/callb.h`, are written to hide `;` in the macro.

And I also found the similar style in your recent commit fe6a7b7 (macro `MEMORY_BARRIER`).

I added those macros simply because the impossibility of using `#if` inside a single `#define`, so I think as long as it looks clear, either style is fine.

Anyway if the other style is preferred, it should still be available on 0ed591168d43bd3cfe7fefea46684f402296bbd0.

**Pull request closed without merging**